### ISSUE TITLE
getting string maps to more than single token

### ISSUE THREAD

**Issue by 1blackbar** [rinongal/textual_inversion#6]
what that means ? should i change it ?No matter what i change it to i get this error over and over ,  should i use two words ? how they should be structured, ? word1,word2 or (word,word) or word1 word2
this is veru important but yet nowhere to be found in instructions
Can you uinclude actual command line to train lets say cartoon t-rex  image by john smith ? I have lot of trouble with this here with SD
Do i need to name my images somehow? my image names are all over the place

**Comment by rinongal** [comment]
This error can come up in one of two places: either your coarse class is not a single token word, or you're using a placeholder string which is not tied to a single token.

What command did you run for training? Did you change anything in the finetuning config file?

**Comment by zoru22** [comment]
I was kind of wondering the same thing. here's my command:

```
python main.py --base configs/latent-diffusion/txt2img-1p4B-finetune.yaml \
-t true \
--actual_resume models/ldm/text2img-large/model.ckpt \
-n leavanny_attempt_one \
--gpus 0, \
--data_root "/home/zoru/Pictures/Pokemons/512/leavannies/" \
--init_word=leavanny
```

Leavanny does not map to a single token.
[leavannies.zip](https://github.com/rinongal/textual_inversion/files/9411113/leavannies.zip)

I ran it for about 200 epochs against the default latent-diffusion model as you can see, then when i try to run for inference, it doesn't get the idea of a leavanny. I think I need to (somehow(?)) update the embeddings.

But it does seem like it gets better at leavanny? Here's the samples -scaled on the 2nd epoch. I don't want to clog this post up with too much irrelevant details but I'm not sure what 5k training iterations means? Is that epochs?
(I realize that I have way more images in the zip than the recommended 3-5 lol)

samples_scaled:
![samples_scaled_gs-006000_e-000002_b-001199](https://user-images.githubusercontent.com/111854061/186284609-ca9670e7-bb9f-4ec5-8905-e9d606ecbe92.jpg)

**Comment by zoru22** [comment]
Apologies for not including more detail in the last comment. I've commented out the assertion on only-one token, and am just hoping the code will do the right thing, it most certainly isn't going to.

**Comment by rinongal** [comment]
I don't recommend disabling the multi-token check. Instead, use an initial word that is a rougher description of the concept, for example you could try "leaf", "drawing" or "painting".  Keep in mind this is not the word that will later represent your concept. It's just the starting point for the optimization.

Unless you changed the config file, the word representing your concept is "*".
So during inference you want to use something like: "a photo of *" or "a * on the beach". Not "a leavanny on the beach".

Regarding number of steps: 
You'll see how the outputs in the log directory (for example the images) have a name of the form: `samples_scaled_gs-002000_e-000029_b-000057`.
The number after _gs is the number of steps (gs stands for global step). You can also see this number in the printouts during training. We used checkpoints from around 5000 steps.

Regarding dataset size: A bigger set is NOT better. I suggest you go over the images in your directory and try to choose about 5 images where the leavanny is alone and more or less front facing.
If it still fails to recreate it, you can try increasing the learning rate (in the config file), this typically makes the shape better but it might make it harder to edit the object later.

If you still have trouble, let me know and I'll try to do a run with your data and see what I can get.

**Comment by zoru22** [comment]
So I did the training run with a few changes

1) Opened `ldm/data/personalized.py` and edited all those prompts and changed from `photo` to `digital illustration`

2) Moved to 1 token: `leaf`

(I have not edited the finetune.yaml file)

Command:
```
python main.py --base configs/latent-diffusion/txt2img-1p4B-finetune.yaml -t true --actual_resume models/ldm/text2img-large/model.ckpt -n leavanny_attempt_two --gpus 0, --data_root "/home/zoru/Pictures/Pokemons/512/leavannies/" --init_word=leaf
```
Images in the leavannies folder:
[5-image-run.zip](https://github.com/rinongal/textual_inversion/files/9421582/5-image-run.zip)

(I did more runs with different combinations)

2) After ~5600 iterations, I stopped it and then ran inference, with the following command:

```
python scripts/txt2img.py --ddim_eta 0 --n_samples 4 --n_iter 2 --scale 10 --ddim_steps 150 --embedding_path /mnt/extended/textual_inversion/logs/leavannies2022-08-24T20-37-04_leavanny_attempt_two/checkpoints/embeddings_gs-5624.pt --ckpt_path /mnt/extended/textual_inversion/models/ldm/text2img-large/model.ckpt --prompt "a painting of a * in the style of russ rutkowski"
```

![a-painting-of-a-*-in-the-style-of-russ-rutkowski](https://user-images.githubusercontent.com/111854061/186582534-13b17c51-2eb9-44fa-8133-ad3c090a2ab8.jpg)

Overall, I'm impressed! The original text2img-large model has _no idea_ what a leavanny is! Still, we have a _bit_ of work to go. I'll post up a fork with my tweaks. I think I'll play with increasing the learning rate a tad.

**Comment by rinongal** [comment]
I managed these which might be a bit better? Unfortunately it's still a way off. Not sure this method can give the level of identity preservation you probably want here, but maybe with SD or with followups :)

![samples_scaled_gs-014500_e-000230_b-000010](https://user-images.githubusercontent.com/19426368/186855834-d5396d12-2e58-4f3d-a533-628ec1ec9be7.jpg)

"A * themed backpack":

![a-%2A-themed-backpack](https://user-images.githubusercontent.com/19426368/186855937-15b577ac-a137-41ed-a920-91090bcca807.jpg)

The embedding:

[leavanny.zip](https://github.com/rinongal/textual_inversion/files/9431077/leavanny.zip)

Either way, it seems like this is resolved so I'm closing the issue. Feel free to reopen if you need more help.

**Comment by zoru22** [comment]
Keep the ticket closed, I'm just circling back to report: I got the Leavanny prompt(s) to "work" with textual inversion/Stable Diffusion (v1-3, not 1-4 or 1-5).


![My favorite Leavanny, though it doesn't _quite_ get where the abdomen underneath "should" be](https://i.imgur.com/EHFQIhm.jpg)

![Oil Painting-style leavanny w/ accompanying dressage](https://i.imgur.com/WUG9EVb.jpg)

Full imgur album here: 

https://imgur.com/a/hQhH9Em

**Comment by liuquande** [comment]
Hi @zoru22 ,

Very excited to see that your results are quite good.

I am trying to increase the quality of my results with the stable diffusion model v 1.3 (in this issue #61 ), would you mind to share the commend and the config file you use, and how many iterations have you trained.

I have use the default SD config file and train the model as follows, but the results are not good (though I think the image I used is not difficult):
``
python main.py --base configs/stable-diffusion/v1-finetune.yaml -t --actual_resume ./models/ldm/stable-diffusion/sd-v1-3.ckpt -n penguin_toy --gpus 0,1,2 --data_root ./img/test/small/ --init_word toy
``

Many thanks!
